WalletWasabi/BitcoinCore/Monitoring/P2pReconnector.cs

```diff
@@ -12,47 +12,41 @@ public class P2pReconnector : PeriodicRunner
 		public P2pReconnector(TimeSpan period, P2pNode p2pNode) : base(period)
 		{
 			P2pNode = Guard.NotNull(nameof(p2pNode), p2pNode);
-			Success = new TaskCompletionSource<object>();
+			Success = new TaskCompletionSource<bool>();
 		}
 
 		public P2pNode P2pNode { get; set; }
-		private TaskCompletionSource<object> Success { get; }
+		private TaskCompletionSource<bool> Success { get; }
 
 		protected override async Task ActionAsync(CancellationToken cancel)
 		{
-			try
+			Logger.LogInfo("Trying to reconnect to P2P...");
+			if (await P2pNode.TryDisconnectAsync(cancel).ConfigureAwait(false))
 			{
-				Logger.LogInfo("Trying to reconnect to P2P...");
-				P2pNode.Disconnect();
 				await P2pNode.ConnectAsync(cancel).ConfigureAwait(false);
-			}
-			catch
-			{
-				P2pNode.Disconnect();
-				throw;
-			}
 
-			Logger.LogInfo("Successfully reconnected to P2P.");
-
-			Success.TrySetResult(null);
+				Logger.LogInfo("Successfully reconnected to P2P.");
+				Success.TrySetResult(true);
+			}
 		}
 
 		public async Task StartAndAwaitReconnectionAsync(CancellationToken cancel)
 		{
 			await StartAsync(cancel).ConfigureAwait(false);
-			using var ctr = cancel.Register(() => Success.SetResult(null));
+			using var ctr = cancel.Register(() => Success.SetResult(false));
```

Comment:
I'm confused because:

* It seems there is only one `P2pReconnector` handles single `P2pNode`.
* Now `private TaskCompletionSource<bool> Success { get; }` is a read-only property.
* `TaskCompletionSource` cannot be set repeatedly. Only once. Otherwise, you'll get an exception. 

but `StartAndAwaitReconnectionAsync` is called repeatedly because P2P node can disconnect multiple times, can't it?

Not sure, whether it's clear what I'm after here. My point is whether it may happen that you set `Success.SetResult()` multiple times or not.

Reply:
Node_DisconnectedAsync can be called once and it won't return until the end of StartAndAwaitReconnectionAsync. If there would be another Node to connect with there would be a problem but there is none.

Reply:
I see, thanks for explanation.